Sming/SmingCore/Network/HttpRequest.cpp

```diff
@@ -82,77 +83,94 @@ String HttpRequest::getContentType()
-	int urlEnd = NetUtils::pbufFindChar(buf, ' ', urlStart);
-	if (urlStart == 0 || urlEnd == -1)
-	{
-		debugf("!BadRequest");
-		return eHPR_Failed;
-	}
-	method = NetUtils::pbufStrCopy(buf, 0, urlStart - 1);
+		int urlStart = tmpbuf.indexOf(" ")+1;
+		urlEnd = tmpbuf.indexOf(" ", urlStart);
+		if (urlStart == 0 || urlEnd == -1)
+		{
+			debugf("!BadRequest");
+			return eHPR_Failed;
+		}
 
-	int urlParamsStart = NetUtils::pbufFindChar(buf, '?', urlStart);
-	if (urlParamsStart != -1 && urlParamsStart < urlEnd)
-	{
-		path = NetUtils::pbufStrCopy(buf, urlStart, urlParamsStart - urlStart);
-		if (requestGetParameters == NULL) requestGetParameters = new HashMap<String, String>();
-		extractParsingItemsList(buf, urlParamsStart + 1, urlEnd, '&', ' ', requestGetParameters);
+		method = tmpbuf.substring(0, urlStart-1);
+		int urlParamsStart = tmpbuf.indexOf("?", urlStart);
+		if (urlParamsStart != -1 && urlParamsStart < urlEnd)
+		{
+			path = tmpbuf.substring(urlStart, urlParamsStart);
+			if (requestGetParameters == NULL) requestGetParameters = new HashMap<String, String>();
+			extractParsingItemsList(tmpbuf, urlParamsStart + 1, urlEnd, '&', ' ', requestGetParameters);
+		}
+		else
+			path = tmpbuf.substring(urlStart, urlEnd);
+		debugf("path=%s", path.c_str());
+		urlEnd = tmpbuf.indexOf("\r\n", urlEnd)+2;
 	}
-	else
-		path = NetUtils::pbufStrCopy(buf, urlStart, urlEnd - urlStart);
-	debugf("path=%s", path.c_str());
 
 	int line, nextLine;
-	line = NetUtils::pbufFindStr(buf, "\r\n", urlEnd)  + 2;
+	line = urlEnd;
 	do
 	{
-		nextLine = NetUtils::pbufFindStr(buf, "\r\n", line);
+
+		nextLine = tmpbuf.indexOf("\r\n", line+2);
 		if (nextLine - line > 2)
+
 		{
-			int delim = NetUtils::pbufFindStr(buf, ":", line);
+			int delim = tmpbuf.indexOf(":", line);
 			if (delim != -1)
 			{
-				String name = NetUtils::pbufStrCopy(buf, line, delim - line);
+				String name = tmpbuf.substring(line, delim);
 				if (server->isHeaderProcessingEnabled(name))
 				{
+					debugf("Name: %s", name.c_str());
 					if (name == "Cookie")
 					{
 						if (cookies == NULL) cookies = new HashMap<String, String>();
-						extractParsingItemsList(buf, delim + 1, nextLine, ';', '\r', cookies);
+						extractParsingItemsList(tmpbuf, delim + 1, nextLine, ';', '\r', cookies);
 					}
 					else
 					{
-						String value = NetUtils::pbufStrCopy(buf, delim + 1, nextLine - (delim + 1));
+						String value = tmpbuf.substring(delim + 1, nextLine);
 						value.trim();
-						if (requestHeaders == NULL) requestHeaders = new HashMap<String, String>();
 						(*requestHeaders)[name] = value;
 						debugf("%s === %s", name.c_str(), value.c_str());
 					}
+
 				}
 			}
+
+		}
+		if (nextLine != -1) {
+			line = nextLine + 2;
 		}
-		line = nextLine + 2;
+
 	} while(nextLine != -1);
 
 	if (headerEnd != -1)
 	{
+		tmpbuf = "";
 		debugf("parsed");
 		return eHPR_Successful;
 	}
+	headerDataProcessed += buf->tot_len;
+	tmpbuf = tmpbuf.substring(line, buf->tot_len);
```

Comment:
In regards to  the size see the talk on substring below. But you caught something else - it should
`tmpbuf.substring(line, tmpbuf.length()` - the length of the tmpbuf can be longer than buf->tot_len

totally missed that